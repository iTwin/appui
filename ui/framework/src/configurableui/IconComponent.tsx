/*---------------------------------------------------------------------------------------------
* Copyright (c) 2018 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
/** @module Item */

import * as React from "react";

/** Properties for the [[Icon]] React component */
export interface IconProps {
  /** CSS class name for icon */
  iconClass?: string;
  /** React element for icon */
  iconElement?: React.ReactNode;
}

/** Icon React component */
export class Icon extends React.Component<IconProps> {
  public render(): React.ReactNode {
    if (this.props.iconClass) {
      const className = "icon " + this.props.iconClass;
      return (
        <i className={className} />
      );
    } else if (this.props.iconElement) {
      return (
        <i className="icon item-svg-icon">
          {this.props.iconElement}
        </i>
      );
    }

    return null;
  }
}
