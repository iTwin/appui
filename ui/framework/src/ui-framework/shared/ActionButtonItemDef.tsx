/*---------------------------------------------------------------------------------------------
* Copyright (c) 2019 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
/** @module Item */

import * as React from "react";

import { CommandHandler, ItemProps } from "./ItemProps";
import { ItemDefBase } from "./ItemDefBase";

import { Size } from "@bentley/ui-ninezone";
import { Orientation } from "@bentley/ui-core";
import { ActionItemButton } from "../toolbar/ActionItemButton";

/** Abstract base class that is used by classes to execute an action when pressed.
 * @public
 */
export abstract class ActionButtonItemDef extends ItemDefBase {
  protected _commandHandler?: CommandHandler;
  public parameters?: any;
  public size?: Size;

  constructor(itemProps: ItemProps) {
    super(itemProps);

    this.execute = this.execute.bind(this);
  }

  public execute(): void {
    if (this._commandHandler && this._commandHandler.execute) {
      if (this._commandHandler.getCommandArgs)
        this._commandHandler.execute(this._commandHandler.getCommandArgs());
      else
        this._commandHandler.execute(this._commandHandler.parameters);
    }
  }

  public handleSizeKnown = (size: Size) => {
    this.size = size;
  }

  public getDimension(orientation: Orientation): number {
    let dimension = 0;
    if (this.size)
      dimension = (orientation === Orientation.Horizontal ? this.size.width : this.size.height) + 1;
    return dimension;
  }

  public toolbarReactNode(_index?: number): React.ReactNode {
    if (!this.isVisible)
      return null;

    return (
      <ActionItemButton
        actionItem={this}
        onSizeKnown={this.handleSizeKnown}
      />
    );
  }
}
