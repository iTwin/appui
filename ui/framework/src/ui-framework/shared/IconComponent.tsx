/*---------------------------------------------------------------------------------------------
* Copyright (c) 2019 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
/** @module Item */

import * as React from "react";

/** Prototype for an IconSpec which can be a string or a ReactNode. */
export type IconSpec = string | React.ReactNode;

/** Properties for the [[Icon]] React component */
export interface IconProps {
  /** CSS class name for icon */
  iconSpec?: IconSpec;
}

/** Icon Functional component */
// tslint:disable-next-line:variable-name
export const Icon: React.FunctionComponent<IconProps> = (props) => {
  if (!props.iconSpec) return null;

  if (typeof props.iconSpec === "string") {
    const className = "icon " + props.iconSpec;
    return (<i className={className} />);
  }
  return (
    <i className="icon item-svg-icon">
      {props.iconSpec}
    </i>
  );
};
