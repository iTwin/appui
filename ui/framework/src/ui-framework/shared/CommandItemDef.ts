/*---------------------------------------------------------------------------------------------
* Copyright (c) 2019 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
/** @module Item */

import { CommandItemProps } from "./ItemProps";
import { ActionButtonItemDef } from "./ActionButtonItemDef";

/** An Item that executes a Command.
 * @public
Â */
export class CommandItemDef extends ActionButtonItemDef {
  private static _sId = 0;
  public static commandIdPrefix = "Command-";
  public commandId: string = "";

  constructor(commandItemProps: CommandItemProps) {
    super(commandItemProps);

    if (commandItemProps.execute) {
      this._commandHandler = { execute: commandItemProps.execute, parameters: commandItemProps.parameters, getCommandArgs: commandItemProps.getCommandArgs };
    }

    if (commandItemProps.commandId)
      this.commandId = commandItemProps.commandId;
    else {
      CommandItemDef._sId++;
      this.commandId = CommandItemDef.commandIdPrefix + CommandItemDef._sId;
    }
  }

  public get id(): string {
    return this.commandId;
  }
}
