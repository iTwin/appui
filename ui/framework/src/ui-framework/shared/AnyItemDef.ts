/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
/** @module Item */

import { GroupItemDef } from "../toolbar/GroupItem";
import { ToolItemDef } from "./ToolItemDef";
import { CommandItemDef } from "./CommandItemDef";
import { ActionButtonItemDef } from "./ActionButtonItemDef";

/** Union of all Item definitions that can be specified in a GroupItem
 * @public
 */
export type AnyItemDef = GroupItemDef | CommandItemDef | ToolItemDef | ActionButtonItemDef;
