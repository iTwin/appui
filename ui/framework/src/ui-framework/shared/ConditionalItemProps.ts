/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
/** @module Item */

import { AnyItemDef } from "./AnyItemDef";
import { ItemProps } from "./ItemProps";

/** Definition for a Conditional item that conditionally renders other items based on UiSync events.
 * @beta
Â */
export interface ConditionalItemProps extends ItemProps {
  conditionalId?: string;
  items: AnyItemDef[];
}
