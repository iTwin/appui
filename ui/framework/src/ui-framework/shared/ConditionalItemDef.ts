/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
/** @packageDocumentation
 * @module Item
 */

import { AbstractConditionalItemProps, OnItemExecutedFunc } from "@bentley/ui-abstract";

import { ItemList, ItemMap } from "./ItemMap";
import { ItemDefBase, BaseItemState } from "./ItemDefBase";
import { SyncUiEventArgs } from "../syncui/SyncUiEventDispatcher";
import { ActionButtonItemDef } from "./ActionButtonItemDef";
import { ConditionalItemProps } from "./ConditionalItemProps";
import { ItemDefFactory } from "./ItemDefFactory";
import { AnyItemDef } from "./AnyItemDef";

/** An Item that conditionally renders other items based on UiSync events.
 * @beta
Â */
export class ConditionalItemDef extends ItemDefBase {
  private static _sId = 0;
  public static conditionalIdPrefix = "Conditional-";
  public conditionalId: string;
  public items: AnyItemDef[];
  private _itemList!: ItemList;
  private _itemMap!: ItemMap;

  constructor(props: ConditionalItemProps) {
    super(props);

    if (props.conditionalId)
      this.conditionalId = props.conditionalId;
    else {
      ConditionalItemDef._sId++;
      this.conditionalId = ConditionalItemDef.conditionalIdPrefix + ConditionalItemDef._sId;
    }

    this.items = props.items;
  }

  /** @internal */
  public static constructFromAbstractItemProps(abstractItemProps: AbstractConditionalItemProps, onItemExecuted?: OnItemExecutedFunc): ConditionalItemDef {
    const items = ItemDefFactory.createItemListForGroupItem(abstractItemProps.items, onItemExecuted);
    const itemProps: ConditionalItemProps = {
      conditionalId: abstractItemProps.conditionalId,
      items,
    };

    return new ConditionalItemDef(itemProps);
  }

  public get id(): string {
    return this.conditionalId;
  }

  public resolveItems(force?: boolean): void {
    if (this._itemList && !force)
      return;

    this._itemList = new ItemList();
    this._itemMap = new ItemMap();

    this.items.forEach((item: AnyItemDef) => {
      const id: string = item.id;
      this._itemList.addItem(item);
      this._itemMap.set(id, item);
    });
  }

  public handleSyncUiEvent(args: SyncUiEventArgs): boolean {
    let refreshState = false;

    this.resolveItems();

    // istanbul ignore else
    if (this.stateSyncIds && this.stateSyncIds.length > 0) {
      refreshState = this.stateSyncIds.some((value: string): boolean => args.eventIds.has(value));

      // istanbul ignore else
      if (refreshState) {
        // istanbul ignore else
        if (this.stateFunc) {
          const oldState: BaseItemState = { isVisible: this.isVisible, isEnabled: this.isEnabled };
          const newState = this.stateFunc(oldState);

          if (newState.isVisible !== undefined && this.isVisible !== newState.isVisible) {
            this.isVisible = newState.isVisible;
            this._updateChildIsVisible(this.isVisible);
          }

          if (newState.isEnabled !== undefined && this.isEnabled !== newState.isEnabled) {
            this.isEnabled = newState.isEnabled;
            this._updateChildIsEnabled(this.isEnabled);
          }
        }
      }
    }

    return refreshState;
  }

  private _updateChildIsVisible(isVisible: boolean) {
    this._itemList.items.forEach((item: ItemDefBase) => {
      item.isVisible = isVisible;
    });
  }

  private _updateChildIsEnabled(isEnabled: boolean) {
    this._itemList.items.forEach((item: ItemDefBase) => {
      item.isEnabled = isEnabled;
    });
  }

  public getVisibleItems(): ActionButtonItemDef[] {
    this.resolveItems();
    const visibleItems = new Array<ActionButtonItemDef>();
    this._itemList.items.forEach((item: ItemDefBase) => {
      if (item.isVisible && item instanceof ActionButtonItemDef)
        visibleItems.push(item);
    });
    return visibleItems;
  }

}
