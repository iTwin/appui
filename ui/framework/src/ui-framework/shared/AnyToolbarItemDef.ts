/*---------------------------------------------------------------------------------------------
* Copyright (c) 2019 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
/** @module Item */

import { ConditionalItemDef } from "./ConditionalItemDef";
import { CustomItemDef } from "./CustomItemDef";
import { AnyItemDef } from "./AnyItemDef";

/** Union of all Item definitions that can be specified in a GroupItem
 * @public
 */
export type AnyToolbarItemDef = AnyItemDef | ConditionalItemDef | CustomItemDef;
