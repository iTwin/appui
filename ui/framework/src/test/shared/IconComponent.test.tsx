/*---------------------------------------------------------------------------------------------
* Copyright (c) 2019 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
import * as React from "react";
import { shallow, mount } from "enzyme";
import { Icon } from "../../ui-framework/shared/IconComponent";

describe("IconComponent", () => {

  it("should render with ReactNode", () => {
    mount(<Icon iconSpec={<span>Test</span>} />);
  });

  it("should render correctly with ReactNode", () => {
    shallow(<Icon iconSpec={<span>Test</span>} />).should.matchSnapshot();
  });

});
