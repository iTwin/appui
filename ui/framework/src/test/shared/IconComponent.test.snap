// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IconComponent should render correctly with ReactNode 1`] = `
<i
  className="icon uifw-item-svg-icon"
>
  <span>
    Test
  </span>
</i>
`;

exports[`IconComponent should render correctly with icon class string 1`] = `
<i
  className="icon icon-developer"
/>
`;

exports[`IconComponent should render correctly with icon src string 1`] = `
<i
  className="icon src:test.svg"
/>
`;
